content/technotes/2019-12-19-urls-checking/index.md

```diff
@@ -0,0 +1,456 @@
+    xml2::xml_find_all("//a") %>%
+    xml2::xml_attr("href") -> urls
+  
+  tibble::tibble(filepath = filepath,
+                 url = urls)
+}
+
+all_urls <- purrr::map_df(mds, get_links)
+all_urls <- all_urls  %>%
+  dplyr::mutate(url = stringr::str_remove_all(url, "#.*"),
+                url = stringr::str_remove(url, "\\/$"))
+
+all_urls
+```
+
+    ## # A tibble: 14,234 x 2
+    ##    filepath                                            url                      
+    ##    <chr>                                               <chr>                    
+    ##  1 /home/maelle/Documents/ropensci/roweb2/content/aut… https://adamhsparks.netl…
+    ##  2 /home/maelle/Documents/ropensci/roweb2/content/aut… https://aldocompagnoni.w…
+    ##  3 /home/maelle/Documents/ropensci/roweb2/content/aut… http://robitalec.ca      
+    ##  4 /home/maelle/Documents/ropensci/roweb2/content/aut… https://alison.rbind.io  
+    ##  5 /home/maelle/Documents/ropensci/roweb2/content/aut… https://dobb.ae          
+    ##  6 /home/maelle/Documents/ropensci/roweb2/content/aut… https://thestudyofthehou…
+    ##  7 /home/maelle/Documents/ropensci/roweb2/content/aut… https://annakrystalli.me 
+    ##  8 /home/maelle/Documents/ropensci/roweb2/content/aut… https://paleantology.com…
+    ##  9 /home/maelle/Documents/ropensci/roweb2/content/aut… https://aurielfournier.g…
+    ## 10 /home/maelle/Documents/ropensci/roweb2/content/aut… https://faculty.washingt…
+    ## # … with 14,224 more rows
+
+We chose a different method to find those within and outside of our
+website.
+
+### Broken internal URLs
+
+When building a Hugo website, one gets a sitemap, which is basically a
+collection of links to all the pages of the website. [If an internal
+link is not in the sitemap, it does not exist.](https://masalmon.eu/404)
+
+We generated the sitemap from within the website folder to extract
+links.
+
+``` r
+cwd <- getwd()
+setwd(website_source)
+
+p <- processx::process$new("hugo", args = "server", echo = TRUE)
+```
+
+    ## Running hugo server
+
+``` r
+Sys.sleep(120)
+
+localhost <- "http://localhost:1313"
+browseURL(localhost)
+paste0(localhost, "/sitemap.xml") %>%
+  xml2::read_xml() %>%
+  xml2::xml_ns_strip() %>%
+  xml2::xml_find_all("//loc") %>%
+  xml2::xml_text() %>%
+  stringr::str_remove_all(localhost) %>%
+  stringr::str_remove("\\/$") -> links
+
+p$kill()
+```
+
+    ## [1] TRUE
+
+``` r
+setwd(cwd)
+head(links)
+```
+
+    ## [1] "/authors/scott-chamberlain"         "/tags/api"                         
+    ## [3] "/authors"                           "/tags/http"                        
+    ## [5] "/technotes/2019/12/11/http-testing" "/tags/mocking"
+
+So these are the existing internal links. We could also have extracted
+them using [the multi-request features of
+`curl`](https://github.com/jeroen/curl/blob/master/examples/sitemap.R).
+
+Let’s now extract the internal links we used in the content.
+
+``` r
+all_urls %>%
+  dplyr::filter(!grepl("^http", url)) -> 
```

Comment:
did we look at ftp links? just a question, not a comment intended to change this post

Reply:
No

Reply:
ok

Reply:
should curl NOT be in backticks?

Reply:
good catch!